Review June 2020

mads-bertelsen opened the pull request:
As this project has not been properly reviewed, it is necessary to make a complete review in order to assure quality.
Due to the size of the task, it is broken down to smaller tasks which are described on Jira to manage the work. This pull request will be used for adding review comments and improvements along the way without affecting the master branch.

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                y_err = data.Error
+
+                #(fig, ax0) = plt.errorbar(x, y, yerr=y_err)
+                plt.errorbar(x, y, yerr=y_err)
```
Then call the methods on the axes, to also be sure they are plotting on the correct axes
```suggestion
                ax0.errorbar(x, y, yerr=y_err)
```

mads-bertelsen replied:
All plotting is now done through the ax object instead of plt directly. Great suggestion.

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                #(fig, ax0) = plt.errorbar(x, y, yerr=y_err)
+                plt.errorbar(x, y, yerr=y_err)
+
+                ax0 = plt.gca()
```
```suggestion
```

mads-bertelsen replied:
Avoided by giving the internal plotting function the ax as an argument.

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                             data.metadata.limits[1]*x_axis_mult)
+
+                # Add a title
+                plt.title(data.metadata.title)
```
```suggestion
                ax0.set_title(data.metadata.title)
```

mads-bertelsen replied:
Done!

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                plt.title(data.metadata.title)
+
+                # Add axis labels
+                plt.xlabel(data.metadata.xlabel)
+                plt.ylabel(data.metadata.ylabel)
```
```suggestion
                ax0.set_xlabel(data.metadata.xlabel)
                ax0.set_ylabel(data.metadata.ylabel)
```

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                                                            max_value)
+
+                # Select colormap
+                cmap = plt.get_cmap('hot')
```
Maybe it would be nice to have the option to change this hard-coded cmap?

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                norm = BoundaryNorm(levels, ncolors=cmap.N, clip=True)
+
+                # Create the figure
+                fig, (ax0) = plt.subplots()
```
This would now be redundant with the `fig` and `ax0` creation above

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                if data.plot_options.custom_xlim_right:
+                    ax0.set_xlim(right=data.plot_options.right_lim)
+
+            elif len(data.metadata.dimension) == 2:
```
I think I would try to split this into 2 different functions, one that plots 1D data, and another that makes 2d images.
Would be more readable and maintainable.

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                fig, (ax0) = plt.subplots()
+
+                # Plot the data on the meshgrids
+                if data.plot_options.log:
+                    color_norm = matplotlib.colors.LogNorm(vmin=min_value,
+                                                           vmax=max_value)
+                    im = ax0.pcolormesh(x, y, to_plot,
+                                        cmap=cmap, norm=color_norm)
+                else:
```
you could make this shorter with something like
```Py
if data.plot_options.log:
    norm = matplotlib.colors.LogNorm(vmin=min_value,
                                                           vmax=max_value)
else:
    norm = BoundaryNorm(levels, ncolors=cmap.N, clip=True)
im = ax0.pcolormesh(x, y, to_plot, cmap=cmap, norm=norm)
```
i don't think `norm` is used anywhere else below, so should be ok to do this

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+                ax0.set_title(data.metadata.title)
+                
+                # Add axis labels
+                plt.xlabel(data.metadata.xlabel)
+                plt.ylabel(data.metadata.ylabel)
```
```suggestion
                ax0.set_xlabel(data.metadata.xlabel)
                ax0.set_ylabel(data.metadata.ylabel)
```

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+        ax = axs.reshape(-1)
+
+        index = -1
+        for data in data_list:
```
You could use `for index, data in enumerate(data_list):`
and then you don't need to initialise or increment `index` manually

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+            print("Plotting data with name "
+                  + data.metadata.component_name)
+
+            if isinstance(data.metadata.dimension, int):
```
So you have an awful lot of duplication here between `make_sub_plot` and `make_plot`.

If `make_sub_plot` is actually doing more or less the same as `make_plot` but displaying on a subplots, it would be much cleaner and easier to maintain/less error prone if `make_sub_plots` just created some subplot axes and then called `make_plot` to populate each subplot.

Since `make_plot` already handles a list of inputs, you could just pass a list of axes as an argument.
Then in `make_plot`, if the axes are not supplied (e.g. if they are `None`), then you create some new ones.
If they are supplied, just use those.

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+        plt.show()
+
+
+class make_sub_plot:
```
I think the recommended convention is to have uppercase letters in class names, and underscores in function names.
e.g. `MakeSubPlot`.

nvaytet replied:
That said, this class only has an `__init__` and no other functions associated to it.
Maybe it can just be a function? Does it need to be a class? Is another class inheriting from it somehere?
If not, I would just make it a function, if it is just making a figure. Matplotlib will take care of keeping the figure alive for you, it doesn't need to be in a class. Especially since I don't really see the use of `self` anywhere in here.

mads-bertelsen replied:
make_plot, make_sub_plot and make_animation are now all functions!

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+        plt.show()
+
+
+class make_animation:
```
In this case this might be useful to have a class, see below

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+        # initialize plots
+        
+        data = data_list[0]
+        if isinstance(data.metadata.dimension, int):
```
Again i would try to just call `make_plot` here instead of duplicating again.

mads-bertelsen replied:
Made a internal plotting function instead that make_plot, make_sub_plot and make_animation all use. The amount of duplication was really bad, glad to be rid of it!

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+
+        def init_1D():
+            # initialize function for animation
+            er.set_data([], [], []) # wont work
```
I would be very careful using functions inside function to access the scope of the `__init__`. I would either
- pass things as arguments (e.g. `er` needs to be an input argument and this function needs to be outside of the class)
- or use the fact that we are inside a class, and make `er` a member of the class (`self.er = ...`) and then make these functions class members, i.e.
```Py
def init_1D(self):
    self.er = set_data([], [], [])
```

mads-bertelsen replied:
The FuncAnimation need a function for updating the data with only an index as argument (unless I am misunderstanding), and in such I need to define a short function inside this function. I find it acceptable since the function is so short.

nvaytet commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+
+class make_sub_plot:
+    """
+    make_plot plots contents of McStasData objects
```
```suggestion
    make_sub_plot plots contents of McStasData objects
```
copied and pasted from above?

mads-bertelsen replied:
Added additional text, but make_plot and make_sub_plot are still close in their description, as the functions are rather similar.

nvaytet reviewed (changes requested):
In summary, there is not much wrong with the plotting code itself, but there is a lot of duplication.
I strongly recommend to make it so that `make_sub_plot` and `make_animation` are just calling `make_plot`.

I would also try to split this in multiple files to make it easier to navigate. You don't have many files in that directory, so it should be fine to so that (i know that when you start getting many many files, it also starts to make things difficult to follow, but you are not there yet).

nvaytet commented on mcstasscript/tests/test_McStasPlotOptions.py:
```diff
@@ -0,0 +1,68 @@
+
+    def test_McStasPlotOptions_default_orders_of_mag(self):
+        """
+        Test that newly created McStasMetaData has correct type
```
These docstrings all the same for all tests. Should be updated.

mads-bertelsen replied:
I have expanded the testing of McStasPlotOptions (only some options were tested), and updated their doc strings.
In addition a new file was added, test_Plotter.py, which adds 600 lines of testing for the plotting functionality, including running the internal plotter function (without plt.show() ) for all 1D/2D and log mode combinations.

This could obviously still be improved in the sense that actual plotting is not tested or visually compared to a known correct plot.

nvaytet reviewed (commented):
I can see that you have tests for the plot options, but can't really see tests on the plotting functions.
I know testing plotting is incredibly difficult, but maybe a first step would be to just call the plotting functions with different parameters, and simply make sure they don't throw any errors.
It would be a start.

mads-bertelsen commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+
+        index = -1
+        for data in data_list:
+            index = index + 1
```
Did a larger refactoring of the plotting functionality, this silly issue was removed. I can't respond directly to the top comment / summary, but the plotting code was refactored to remove the code duplication. The new version has been pushed to the master branch. The three classes are now all functions, which depend on a simpler underlying function. This also allowed easier and better testing!

mads-bertelsen commented on mcstasscript/interface/plotter.py:
```diff
@@ -0,0 +1,650 @@
+
+            print("Plotting data with name " + data.metadata.component_name)
+            if type(data.metadata.dimension) == int:
+                fig = plt.figure(0)
```
This change was included in the refactor, the internal plotter function now uses fig and ax in the arguments which is provided by the calling function. In that way plt.gca is never used, and the ax object is used for all plotting.

figsize is now also a keyword argument in all three plotting functions.

Commit:
Adding axis1_values and axis2_values to view which plots lines
at the indicated value (lists) to make it easy to check things
in views are as calculated.

Fixed bug in instrument diagram target_index.

Made all right side arrows on instrument diagram in try blocks
as they are not critical. Prints a message if they fail.

g5t committed:
[Fix] Determine base for C integer literals

mads-bertelsen committed:
Merge pull request #59 from g5t/85-base-10-int

[Fix] Determine base for C integer literals

Commit:
Small fix of component reader when finding vectors

Commit:
Update syntax for plotting

Commit:
Update with McStas 3.x relevant syntax

Commit:
Apply McStas 3.x syntax

Commit:
Update syntax to get plot outputs

Commit:
Add Union_init and  Union_stop for 3.x support

Commit:
Updates for Union in McStas 3.x

Commit:
Update that removes the need to provide a name in often used methods
provided the resulting object is stored in a python variable whos
name can be reused. For example

wavelength = instrument.add_parameter("wavelength", value=5)

wavelength = instrument.add_parameter(value=5)

Are eqvivalent.

The methods with this feature are:
    add_parameter
    add_declare_var
    add_user_var
    add_component
    copy_component

Commit:
Addition of ipympl as dependency to fix issues with jupyter notebook widgets not displaying.
Avoid running checks for errors when settings(checks=False) is used, was used in diagrams.

mads-bertelsen committed:
Merge pull request #63 from PaNOSC-ViNYL/implicit_naming

Inferred names

Commit:
Update of actions to new setup-python version and enabling manual dispatch.

Commit:
Update of syntax

Commit:
Issue with syntax in github actions.

Commit:
Updated duplicate name.

mads-bertelsen committed:
Merge pull request #64 from PaNOSC-ViNYL/actions_update

Update of actions and allow manual start of actions

Commit:
Update of integration tests to support McStas 2 and 3.

Include astrox folder in component_reader so McXtrace can find those components.

Commit:
Added openacc and NeXus option to settings, including unit tests.

Updated version to 0.0.57.

Small corrections to docs, including showing McStas 2 / 3 flexibility in Union intro.

Commit:
Added search and dependency keywords.

set_dependency on instrument, reset with set_dependency("").

add_search on instrument (handled with a SearchStatement and SearchStatementList class)
add_search on component

In a search statement it is possible to set SHELL to True as keyword argument to get that feature.

Both component and instrument has:
show_search to display current search statements
clear_search to reset the list to empty

Commit:
Update of documentation and fix of bug in resetting dependency.

Commit:
reset notebook.

mads-bertelsen committed:
Merge pull request #62 from McStasMcXtrace/master

A few notebook edits for compatibility with McStas 3.x

Commit:
Small changes to readme.

Commit:
Updating checks of c variable names with regular expressions and
list of reserved keywords.

Commit:
Updated version of component reader, now handles comments in parameter region.

Commit:
Fixed bugs that caused infinite loops.

Commit:
Fixed spelling mistake in comment.

Commit:
Update version number as behavior changed.

mads-bertelsen committed:
Merge pull request #69 from PaNOSC-ViNYL/comp_reader_improvements

Component reader improvements, now ignores multiline comments

g5t committed:
[Apply][Fix] On top of the up-to-date master

g5t committed:
[Fix] white space issues for tests

Commit:
Small bug pointed out by Kristine which prevented diagnostic output
when the instrument used target_index logic.

Commit:
Small updates:

Added zlabel to colorbar in plotting and metadata
tight_layout for figures written to disk
small syntax fixes.

Commit:
Update that allows reading of nexus files with load functions.

backengine will return read nexus data in usual format.

Commit:
Avoid plotting event data as the results are not meaningful and often crash the plotter.

Patched issue in instrument diagrams for Union systems that use a variable or parameter
for the number of activations, as this couldn't be directly read. For now this simply
assumes a value of 1.

Bumped version for new release.

mads-bertelsen committed:
Merge pull request #73 from PaNOSC-ViNYL/nexus_reader

Nexus support

Updates to the file reader so they now handle NeXus data as well as text data. Before McStasScript would throw an error when trying to read the written nexus file, even when it was successfully written. Now the file is read and returned in the usual format of McStasDataBinned and McStasDataEvent objects.

Also a few other updates:

- Plotting now skips event data as they are not meaningful to plot and often caused issues.
- Diagrams with variables or parameters used for the number_of_activations in Union geometries can now be shown, though it is assumed their number_of_activation counter is 1.
- Removed keyword arguments from run_simulation in managed_mcrun, this should all be handled through the settings dictionary which is cleaner than having two ways of changing the same parameters.

Commit:
Added h5py to requirements to ensure users have the package for nexus reading.

Commit:
Fixed issue with placement of plot in widget on newer ipywidget versions.

mads-bertelsen committed:
Merge pull request #75 from PaNOSC-ViNYL/new_widget_layout

Fixed issue with placement of plot in widget on newer ipywidget

Commit:
mcdisplay is now searched for in both exectuable path and package path.

fixed scaling of plotting when there are many subplots.

avoided plots appearing during unittests.

Commit:
Tight layout on all figures

Commit:
Bump of version.

Commit:
Fixed called_with parameters in unittests for plotting.

mads-bertelsen committed:
Merge pull request #76 from PaNOSC-ViNYL/show_instrument_fix

Mcdisplay issue and plotting fix

Commit:
Hotfix of mcdisplay path issue.

Commit:
Hotfix of mcdisplay path issue.

Commit:
Improved reliability of reading NeXus data.

Removed intensity from event data, as it was only used in plotting, and now plotting of event data has been removed.

mads-bertelsen committed:
Merge pull request #77 from PaNOSC-ViNYL/nexus_loading_improvements

Improved reliability of reading NeXus data

Commit:
Mocked datetime to avoid a flaky test when writing instrument code.

willend committed:
Implement simple solution to auto-detect McStas and McXtrace parameters if run from within a related 'environment'

mads-bertelsen committed:
Merge pull request #78 from willend/master

Simple solution to auto-detect McStas and McXtrace config

Commit:
Fix of case where MCSTAS was not found in environ.

willend committed:
Allow McXtrace sessions to visualize using mxdisplay(-webgl/-pyqtgraph)

willend committed:
We should of course still use mcdisplay in the McStas case...

mads-bertelsen committed:
Merge pull request #80 from willend/master

Allow McXtrace sessions to visualize using mxdisplay(-webgl/-pyqtgraph)

mads-bertelsen committed:
Increasing python version for workflows

mads-bertelsen committed:
Added instrument checker.

mads-bertelsen committed:
Updates to workflows.

mads-bertelsen committed:
Changed how python versions are specified

mads-bertelsen committed:
Shifted to libpyvinyl branch without Pint.

mads-bertelsen committed:
Further development of instrument_checker and version bump.

mads-bertelsen committed:
Updated search statements so McStasScript loads the appropriate folder of components.

When search statements are cleared, the component database is reset.

mads-bertelsen committed:
Added system that writes down all final values given to components for both parameters, AT and ROTATED.

The system can be used for debugging, may be used for some automatic functions in the future.

mads-bertelsen committed:
Merge pull request #83 from PaNOSC-ViNYL/include_meta

Write component parameters

nvaytet committed:
fix bad escape sequence warning in python 3.11

mads-bertelsen committed:
Merge pull request #84 from nvaytet/fix-py-311

Fix bad escape sequence warning in python 3.11

mads-bertelsen committed:
Version bump after merging pull request on python 3.11 compatability.

mads-bertelsen committed:
Fixed issues with escape characters.

mads-bertelsen committed:
Get instrument search object to be written before the first component instead of near DEFINE and parameters. This syntax works where the others have issues.

mads-bertelsen committed:
Bump version for SEARCH fix

mads-bertelsen committed:
Updated test for new SEARCH location.

mads-bertelsen committed:
Merge pull request #85 from PaNOSC-ViNYL/search_fix

Fixing SEARCH system

MilanKlausz committed:
fixed typos in Union tutorial docs

MilanKlausz committed:
fixed typos in Union tutorials

mads-bertelsen committed:
Fixed bug where dependency line was read as additional parameters in components.

mads-bertelsen committed:
Included component made for unit test that lacks these sections and has dependency.

mads-bertelsen committed:
Added handling of damaged utf-8 strings and EQ for search statements.

mads-bertelsen committed:
Change to how NeXus data is read.

Before the nexus file would be opened and closed for each monitor
to allow using the same code when either one or all monitors were
loaded. This may have caused issues on VISA.

Now the file is opened once and all data read when the user wants
all monitors, and it is handdled differently when just one monitor
is loaded, leading to slighlty more complex code, but hopefully
fix the issue with mulitple readers on VISA.

mads-bertelsen committed:
Merge pull request #91 from PaNOSC-ViNYL/nexus_improvement

Change to how NeXus data is read.

willend committed:
Let McStasScript default to use of webgl-classic vis.
Will function from notebooks, provided McStas 3.5.35 is used.

mads-bertelsen committed:
Merge pull request #93 from willend/master

Let McStasScript default to use of webgl-classic visualization.

mads-bertelsen committed:
Update default instrument visualization to webgl-classic

willend committed:
Add ':' consistently in Instrument: name header fields, prompt user to fill in meaningful information.

willend committed:
Add a missing newline, push parm table below %Parameters

willend committed:
Partial generation of parameter list

willend committed:
Add missing +

willend committed:
Add param.comment as parameter description

willend committed:
Drop (type) since this is part of the comment string already.

mads-bertelsen committed:
Merge pull request #86 from MilanKlausz/master

Fix typos in docs and tutorials, Thanks Milan!

mads-bertelsen committed:
Merge pull request #95 from willend/master

Increase compatibility with mcdoc util

mads-bertelsen committed:
Merge branch 'master' into fix-format-strings-on-master

mads-bertelsen committed:
Merge pull request #70 from g5t/fix-format-strings-on-master

Update string format representation used in instrument file output

mads-bertelsen committed:
Updated tests after merging recent pull requests.
Added default value to _read_configuration so a line_length is always available.

Added pypi publish ci

mads-bertelsen committed:
Updated manifest to include test_input_folder in tests

mads-bertelsen committed:
Updated version

willend committed:
Allow use of mcrun --showcfg=resourcedir as fallback for
missing $MCSTAS (and similar for McXtrace)

willend committed:
Add pull_request: under on: in unit_test workflows

mads-bertelsen committed:
Merge pull request #99 from willend/master

Allow use of mcrun --showcfg=resourcedir as fallback for missing $MCSTAS

dependabot[bot] committed:
Bump pypa/gh-action-pypi-publish in /.github/workflows

Bumps [pypa/gh-action-pypi-publish](https://github.com/pypa/gh-action-pypi-publish) from 1.12.4 to 1.13.0.
- [Release notes](https://github.com/pypa/gh-action-pypi-publish/releases)
- [Commits](https://github.com/pypa/gh-action-pypi-publish/compare/v1.12.4...v1.13.0)

---
updated-dependencies:
- dependency-name: pypa/gh-action-pypi-publish
  dependency-version: 1.13.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

mads-bertelsen committed:
Merge pull request #97 from PaNOSC-ViNYL/dependabot/github_actions/dot-github/workflows/pypa/gh-action-pypi-publish-1.13.0

Bump pypa/gh-action-pypi-publish from 1.12.4 to 1.13.0 in /.github/workflows

mads-bertelsen committed:
Adds keyword argument to return figure, ax objects and fix formatting problem in exponential notation on plots

mads-bertelsen committed:
version bump

mads-bertelsen committed:
Merge pull request #100 from PaNOSC-ViNYL/return_fig_ax

Allows make_plot and make_sub_plot to return fig and ax object

willend committed:
First steps toward solving https://github.com/PaNOSC-ViNYL/McStasScript/issues/101

mads-bertelsen committed:
Fixed case where no intensity reported in diagram cause it to fail.
Fixed plotting of 0D monitors as metadata had changed field name.

mads-bertelsen committed:
Merge pull request #103 from willend/master

First steps toward solving #101

mads-bertelsen committed:
Merge pull request #106 from PaNOSC-ViNYL/diagram_updates

Crash fixes for diagram and plotting

0D files can now be plotted again and one failure of diagrams with analysis mode addressed.

mads-bertelsen committed:
Fixed issue #60 by avoiding the error and setting the key as _N instead of overwriting with additional MCPL keys

mads-bertelsen committed:
Added logic to select either restore_neutron or restore_xray as appropriate when making diagnostics instruments.

mads-bertelsen committed:
copy in McStasScript now follows typical McStas logic, as it does not carry keywords:

WHEN
GROUP
JUMP
SPLIT

comments are also removed and serch statement reset.

mads-bertelsen committed:
Fixed issue where instrument names with Param would cause a read failure.

mads-bertelsen committed:
Fixed issue of finding MCPL output when not using a literal string
as the variable input. Now also finds it if its a parameter or
a char array.

mads-bertelsen committed:
Bump version for release.

mads-bertelsen committed:
Merge pull request #107 from PaNOSC-ViNYL/maintenance

Maintenance, bug fixes

Resolves #60, several MCPL files work and get sensible keys in the libpyvinyl dictionary

Resolves #79, now looks for whether restore_neutron or restore_xray parameters are available on the monitor used and use the appropriate one

Resolves #92, now copy doesn't grab EXTEND, SPLIT, GROUP or WHEN. Still considering if component location should be copied, think that would make the most sense

Resolves #96, instruments with "Param" in their name will now have data loaded, checked for other similar possible issues but seems to be the one word not allowed

Resolves #102, every time an instrument simulation is performed, a textfile called component_parameters.txt is written by the mcstas simulation that contains raw values for component parameters, this is used to get the actual filename and not just the variable name.

mads-bertelsen committed:
Fixing issue with component reader, now reads all components in recent McStas.

Small improvements to save_pars systems, being careful not to cause a crash.

Fix of instance where diagram in analysis mode fails without good feedack, improved.

mads-bertelsen committed:
Merge pull request #108 from PaNOSC-ViNYL/bug_fixing

Incremental improvements

sasmodels included in component locations, component reader updated for newer mcstas keywords, save_pars crash fixed and more graceful diagram analysis failure.

mads-bertelsen committed:
Initial version of geometry viewer that runs in python instead of IFrame.

Still not functional or called from interface classes.

mads-bertelsen committed:
Revert "Initial version of geometry viewer that runs in python instead of IFrame."

This reverts commit 40ea85484c3c87f74df85c4c77d5f4695bc57e7b.

This commit was pushed to master instead of a seperate branch, being reverted until ready for production.

mads-bertelsen committed:
Updates to diagrams and diagnostics, more resiliant to input of different types.

mads-bertelsen committed:
Version bump

mads-bertelsen committed:
Merge pull request #109 from PaNOSC-ViNYL/diagram_target_fix

Diagrams had some issues that would crash them or print annoying output in certain cases. Thanks Celine for catching this.

Fixed case where:

Parameter given with a value identical to the default, making the diagram believe a meaningful value is present while McStas would ignore it

Cases where variables are given of a different type than expected, especially taget_index that could have value such as "+ 2" where the space would trip up the conversion to integer

Improved error message when diagram fails to run simulation.

mads-bertelsen committed:
Prototype of overwriting set_parameters from libpyvinyl with version
that knows available parameters when user asks for help

mads-bertelsen committed:
Updated literal string help to support python 3.9

willend committed:
Allow McStasScript to support mpi='auto'
(and let mcrun run with max. parallelisation allowed by OS.)

mads-bertelsen committed:
Merge pull request #114 from willend/master

Allow McStasScript to support mpi='auto'

Still todo: Add a test case, fix widget input

mads-bertelsen committed:
Update adressing reviewer concerns on type matching and raw dict input being usable alongside kwargs

mads-bertelsen committed:
First version of help system for set_parameters on component objects.

mads-bertelsen committed:
Avoid modifying input dicts in set_parameters.

mads-bertelsen committed:
Fixed issue with show_settings failing when using mpi=auto
Allowed mpi=auto in widget
Added tests for mpi=auto

mads-bertelsen committed:
Merge pull request #115 from PaNOSC-ViNYL/widget_mpi_auto

Updates related to mpi=auto

mads-bertelsen committed:
Handling of mcrun return code.

mads-bertelsen committed:
Merge pull request #110 from PaNOSC-ViNYL/parameter_feedback

System for providing help on set_parameter methods on instrument and component objects.

Overwrites set_parameters on instrument object originally inherited from libpyvinyl

mads-bertelsen committed:
Update to tests and print output

mads-bertelsen committed:
Merge branch 'master' into return_code_handling

mads-bertelsen committed:
Merge pull request #116 from PaNOSC-ViNYL/return_code_handling

Handling of mcrun return code

mads-bertelsen committed:
Hotfix of event_plotter and dataformat

mads-bertelsen committed:
Version bump

mads-bertelsen committed:
semibold as a specifier of fontweight is getting deprecated, using value instead.

mads-bertelsen committed:
Merge pull request #117 from PaNOSC-ViNYL/fontweight

semibold as a specifier of fontweight is getting deprecated, using value instead.

mads-bertelsen committed:
Improvement to geometry viewer, now somewhat functional.
Supports multiline, circle, cylinder and polyhedra

Still needs the json to be generated with mcdisplay-webgl --nobrowse call manually

mads-bertelsen committed:
Further progress on instrument visualization.

Now circles are shown from each side, not just one.
Cones now work.
Transformations of both component rotation and align axis given for the same shape provide a meaninful transformation.

Still lack a few defined geometries.
Will need some more control of material given need, the solid geometries should probably be more opaque by default.

mads-bertelsen committed:
Fixed polyhedra. Found issue with rotated arms, cones not where they should be.

mads-bertelsen committed:
Updates to instrument visualization with pythreejs.

Performs well now that lines are merged and fewer objects are made.
Materials are cached to avoid recreating the same material.

mads-bertelsen committed:
Fixed bug that meant some shapes were duplicated.

Now generates the json file using mcdisplay-webgl --nobrowse if not given.

Now has widget for navigation, but for some reason the first move takes
a long time, after that its fast.

mads-bertelsen committed:
Few fixes of mcdisplay_runner in case of using output path and input path

Ensure shape is set to None in loop reading shapes.

mads-bertelsen committed:
Starting work on code that guess how to plot a component based on parameters

mads-bertelsen-agentic committed:
Add tests for instrument_diagnostics module

76 new tests covering View, DiagnosticsPoint, DiagnosticsInstrument,
BeamDiagnostics, EventPlotter, PlotOverview, IntensityDiagnostics, and
common_range_limits. Also adds dummy Monitor_nD component for monitor
tests and updates ComponentReader test counts accordingly.

mads-bertelsen-agentic committed:
Address PR review comments: bump version, add Agg backend comment, add isinstance checks

mads-bertelsen-agentic committed:
Add AGENTS.md with project overview and contribution notes

mads-bertelsen committed:
Merge pull request #1 from mads-bertelsen-agentic/tests/instrument-diagnostics

Add tests for instrument_diagnostics module and basic AGENTS.md

mads-bertelsen committed:
Merge pull request #118 from mads-bertelsen-agentic/master

First agentic work - additional tests and AGENTS.md

mads-bertelsen-agentic committed:
Fix duplicate GHA triggers on PRs and ignore test artifacts

- Restrict push trigger to master only in all 3 unit test workflows,
  preventing duplicate runs when pushing to PR branches
- Add test_instrument.instr to .gitignore (test-generated artifact)
- Add LOCAL_AGENTS.md to .gitignore

mads-bertelsen-agentic committed:
Add integration tests workflow triggered on PRs

Installs McStas from conda-forge, then overlays the PR branch
version of McStasScript on top to test against latest McStas.

mads-bertelsen committed:
Merge pull request #119 from mads-bertelsen-agentic/fix/duplicate-gha-pr-triggers

Fix duplicate GHA triggers on PRs and ignore test artifacts

mads-bertelsen-agentic committed:
Fix: use simple needs reference for matrix job dependency

mads-bertelsen-agentic committed:
Fix flaky timestamp tests by mocking datetime

Both test_write_full_instrument_dependency and
test_write_full_instrument_search now mock datetime.datetime
to avoid race conditions between write and assertion.

mads-bertelsen-agentic committed:
Merge pull request #3 from mads-bertelsen-agentic/add/integration-tests-gha

Add/integration tests gha - installs mcstas from conda-forge and overwrites McStasScript install

mads-bertelsen-agentic committed:
Revert to pull_request only to avoid duplicate runs

Push to new branches without a PR won't trigger tests, but this
avoids duplicate runs (including integration tests) when pushing to
a branch with an open PR.

mads-bertelsen committed:
Merge pull request #120 from mads-bertelsen-agentic/master

Integration tests that install mcstas from conda-forge and overwrites mcstasscript with the tested branch

mads-bertelsen-agentic committed:
Streamline docs build: micromamba env, remove hardcoded McStas paths, add --execute flag

- Add docs/environment.yml for micromamba doc build environment
- Add docs/build_docs.sh with --execute flag for notebook execution (300s timeout)
- Remove hardcoded McStas paths from tutorial and installation notebooks
- Update installation notebook to note conda auto-configuration
- Add docs/source/_autosummary/ and docs/build/ to .gitignore
- Fix docs/requirements.txt, conf.py, Makefile, index.rst

mads-bertelsen committed:
Updating documentation notebooks for mcstas 3.x

mads-bertelsen-agentic committed:
Merge pull request #5 from mads-bertelsen-agentic/clean-docs-workflow

Streamline docs build and updated documentation to McStas 3

mads-bertelsen-agentic committed:
Add GitHub Action to build docs on tag or manual trigger

mads-bertelsen-agentic committed:
Install McStas via micromamba in docs CI workflow

mads-bertelsen-agentic committed:
Merge pull request #6 from mads-bertelsen-agentic/clean-docs-workflow

Add GitHub Action to build docs on tag or manual trigger

mads-bertelsen-agentic committed:
Deploy docs to GitHub Pages on tag push

mads-bertelsen-agentic committed:
Merge pull request #7 from mads-bertelsen-agentic/clean-docs-workflow

Deploy docs to GitHub Pages on tag push

mads-bertelsen committed:
Merge pull request #122 from mads-bertelsen-agentic/master

Documentation overhaul

Addresses several issues with the documentation

- outdated notebooks that didn't run with mcstas 3
- mentions mcstas-pygen in section for instrument reader
- now use mpi="auto" to speed up building documentation
- less manual deployment, can be built locally with an sh script
- GitHub action that can build documentation and deploy on new tags
- Documentation build action can also be executed manually

mads-bertelsen-agentic committed:
Add retry on notebook execution failure

mads-bertelsen-agentic committed:
Remove libpyvinyl note from rendered docs

mads-bertelsen-agentic committed:
Merge branch 'PaNOSC-ViNYL:master' into master

mads-bertelsen-agentic committed:
Update version history documentation (0.0.59–0.0.85)

mads-bertelsen committed:
Update version_history.ipynb

Small corrections to version history

mads-bertelsen committed:
Merge pull request #123 from mads-bertelsen-agentic/master

Documentation updates

mads-bertelsen-agentic committed:
Update README: installation, configuration, and method overview

mads-bertelsen committed:
Update README.md

Updated README, removes outdated information and match current methods on the object types.

mads-bertelsen committed:
Merge pull request #8 from mads-bertelsen-agentic/updated_README

Update README: installation, configuration, and method overview

mads-bertelsen committed:
Update README.md

mads-bertelsen committed:
Merge pull request #124 from mads-bertelsen-agentic/master

Update of README

mads-bertelsen committed:
Move complexity from viewer to mcdisplay_runner

mads-bertelsen committed:
Avoid slow loop and use existing get_component method.

mads-bertelsen-agentic committed:
geometry_viewer: move json_dict init into InstrumentModel constructor

mads-bertelsen-agentic committed:
geometry_viewer: refactor drawcall parsing with dispatch registry

mads-bertelsen-agentic committed:
geometry_viewer: update TODO with completed items

mads-bertelsen-agentic committed:
Refactor geometry_viewer: decouple shapes from rendering backends

- Shapes are now pure dataclasses in model/shapes.py (no pythreejs imports)
- Transform is pure numpy data with transform_points() method
- RendererBackend ABC with render_shape(), apply_transform(), create_material(), make_scene()
- PyThreejsRenderer extracted from old shapes.py + pythree_specific.py
- MatplotlibRenderer with 3D (Poly3DCollection) and 2D projection modes
- LineDescriptor dataclass defers line drawing to make_scene (no circular axis dependency)
- API: view(), view_with_json(), view_with_guess() accept backend parameter
- Config centralized in config.py (colors, segments, camera defaults)
- Fixed bugs: compnent_list typo, undefined transform, return RuntimeError, guess-first fallback
- Proper __init__.py with full public API exports
- All 437 existing tests pass

mads-bertelsen-agentic committed:
Fix matplotlib circle rendering: operator precedence in index wrap

mads-bertelsen-agentic committed:
Add unit tests for geometry_viewer and fix two bugs discovered by tests

- 69 tests covering transforms, shapes, drawcall parsers, ComponentModel,
  InstrumentModel, config, and API backend routing
- Fix specified_pars inverted condition: == should be != in
  guess_geometry_from_comp_object
- Fix check_conditions operator precedence: add parentheses around 'in'
  checks to avoid Python chained comparison bug
- Add docstrings to all test methods
- Copy instrument.json fixture from mcdisplay output

mads-bertelsen-agentic committed:
geometry_viewer: consolidate refactor, replace plan docs with summary

- Remove REFACTOR_PLAN.md, COMBINE_VISUALIZERS.md, TODO.md (all goals implemented)
- Add summary.md with remaining considerations
- Consolidate mcdisplay execution into shared helpers in mcdisplay.py
- Thin show_instrument() wrapper delegates to geometry_viewer.view()
- Unified view() with backend routing for all visualization backends

mads-bertelsen-agentic committed:
geometry_viewer: lazy pythreejs import, webgl Vite server fix, pythreejs availability check

- Make PyThreejsRenderer import lazy so module loads without pythreejs
- show_instrument() checks for pythreejs and raises ImportError with install hint
- webgl backend: don't set nobrowse so mcdisplay opens browser with correct Vite URL
- Print informative message in notebook before starting mcdisplay-webgl
- Fix PyThreejsRenderer NameError in view_with_json()
- Update summary.md with current state

mads-bertelsen-agentic committed:
ci: install pythreejs in GitHub Actions for tests

mads-bertelsen-agentic committed:
instr: guard shutil.which(None) in _read_calibration for McStas and McXtrace

When MCSTAS/MCXTRACE env var is set but mcrun/mxrun is not on PATH,
shutil.which returns None and os.path.dirname(None) crashes. Now falls
back to config file. Also guards the else branch (same end result,
different exception path).

mads-bertelsen-agentic committed:
transform: add from __future__ import annotations for Python 3.8/3.9 compat

The X | None union syntax requires Python 3.10+. The __future__ import
makes annotations lazy strings, allowing the syntax on all versions.

mads-bertelsen-agentic committed:
geometry_viewer: add from __future__ import annotations for Python 3.8/3.9 compat

The X | None union syntax in api.py, model/shapes.py, and
renderer/matplotlib.py requires Python 3.10+. The __future__ import
defers annotation evaluation, allowing the syntax on all versions.

mads-bertelsen-agentic committed:
shapes: remove dataclass kw_only for Python 3.8/3.9 compat

kw_only=True requires Python 3.10+. Made Shape a plain ABC and
defined transform/style fields on each concrete shape class instead
of inheriting from a dataclass parent.

mads-bertelsen-agentic committed:
renderer/base: add from __future__ import annotations for Python 3.8/3.9 compat

list[Any] in type annotations requires Python 3.9+ (or __future__ on 3.8).

mads-bertelsen-agentic committed:
Fix matplotlib backend: orientation, dual output, and aspect ratio

- quaternion_to_rotation_matrix() in transform.py for converting quaternions to rotation matrices
- _transform_points() now uses final_quaternion() to apply combined rotation (rotation_matrix + quaternion), fixing circle/cylinder/cone orientation
- _sample_cylinder_mesh() generates Y-aligned geometry to match pythreejs and default_axis convention
- _remap_to_display() remaps data (X,Y,Z) → display (Z,X,Y) so Z is horizontal
- view_with_json/view_with_guess call plt.show() and return None for matplotlib backends, suppressing duplicate static PNG in notebooks
- set_box_aspect with data extents for proper aspect ratio
- view_init(elev=0, azim=-90) for side view along beam direction
- 6 new tests (75 total)

mads-bertelsen-agentic committed:
Add projection parameter to matplotlib_2d backend

mads-bertelsen-agentic committed:
Fix pythreejs make_scene **kwargs and matplotlib component colormode

mads-bertelsen-agentic committed:
Add colormode dropdown selector to pythreejs widget

mads-bertelsen-agentic committed:
Document colormode dropdown in summary

mads-bertelsen-agentic committed:
Add intensity-based component coloring with colorbar

- Add 'intensity' colormode to color components by simulation data
- Add view_with_analysis() that auto-runs IntensityDiagnostics and visualizes
- Support aggregation modes: total, min, max, mean/average, median, span
- Aggregations operate on axis values (e.g. wavelength) weighted by intensity
- Add colorbar beside scene for component and intensity colormodes
- Auto-detect colorbar label from monitor metadata (xlabel or 'Intensity [n/s]')
- Colorbar updates live when switching colormodes in pythreejs
- Matplotlib backend gets native fig.colorbar() for 3D and 2D modes
- Add 117 unit tests (all passing)

Commit:
feat: integrate on-demand intensity simulation into geometry viewer widget

Merge view_with_analysis() into show_instrument() flow. The pythreejs
widget now always offers 'Intensity' colormode, running the simulation
on-demand when the user selects it.

- Add intensity controls: ncount (IntText), variable dropdown, Limits
  checkbox with Min/Max fields, aggregate dropdown, Run button
- Grey out components when data is stale (variable/ncount/limits changed)
- Cache simulation data so aggregate changes don't require re-run
- Disable controls and show hourglass during simulation
- Pass instrument_object through renderer for simulation access
- Update tests: 562 passing

Commit:
Fix colorbar: use embedded ipympl figure with native matplotlib tick locators

Replace PNG-based colorbar with a live ipympl embedded matplotlib
figure in the PyThreejsRenderer. Uses LogLocator/LogFormatterMathtext
for log-scale and MaxNLocator for linear-scale ticks. Single persistent
figure/axes is reused across colormode updates via ax.cla(). Also
improved matplotlib backend colorbar sizing and label formatting.

Commit:
feat: extend geometry viewer component controls

Commit:
Add geometry viewer guess diagnostics

Commit:
Remove temporary Markdown notes

Commit:
Update geometry_viewer module: add bounds and style models, refactor renderers

Commit:
Improve geometry viewer intensity controls

Commit:
Fix notebook diagram canvas rendering

Commit:
Add intensity orders of magnitude control

Commit:
Fix intensity diagnostic ncount setting

Commit:
Improve geometry viewer usability

Commit:
Document GitHub CLI authentication

mads-bertelsen-agentic committed:
Move GitHub CLI guidance to local notes

mads-bertelsen-agentic committed:
Fix geometry viewer imports on Python 3.9

mads-bertelsen-agentic committed:
Address geometry viewer review feedback

mads-bertelsen committed:
Merge pull request #9 from mads-bertelsen-agentic/geometry_viewer_widget

Geometry viewer widget

mads-bertelsen committed:
Merge pull request #125 from mads-bertelsen-agentic/master

On the agentic fork there has been a lot of work on a new geometry visualization system, originally inspired by Neils efforts here: https://github.com/nvaytet/mcstas-instrument-viewer

This version has choice of visualizer backends

mcdisplay-webgl (in new browser tab)
mcdisplay-webgl-classic (in notebook)
pyqt backend (in window)
pythreejs (new, in notebook)
matplotlib (new, in notebook)
matplotlib_2d (new, in notebook)
The new ones can be used in guess mode where geometry is guessed without having to compile the McStas instrument, but it takes an explicit keyword argument.

The pythreejs version can run diagnostics simulations showing intensity distribution within the instrument, more updates on this area will come, but it is functional. It can also show information by clicking on components.

Both the new backends support coloring components after their position in the file which is helpful for debugging.

The main thing is that it's less browser dependent than the old IFrame solution, and the visualization is of reasonably good quality with pythreejs. It is not added as a dependency yet, so users are warned if they ask for it while not installed. The matplotlib backend is great for the 2d views, but also acts as a fallback in case pythreejs becomes problematic as a dependency.

Comprehensive test coverage of the new features are included.

Review conducted on the mads-bertelsen-agentic fork.

Commit:
Fix geometry viewer backends and diagnostics

Commit:
Use pythreejs for guessed geometry

Commit:
Remove unrelated simulation interface changes

mads-bertelsen committed:
Merge pull request #126 from mads-bertelsen-agentic/geometry_fixes

Updates behavior of output_path and input_path with mcdisplay wrapper. Adds support for index_min in guess mode. Better error handling. Default viewer is back to webgl-classic, but the new backends are available.

Commit:
version bump

mads-bertelsen-agentic committed:
Enable pythreejs in documentation builds

mads-bertelsen-agentic committed:
Document webgl-classic as default backend

mads-bertelsen-agentic committed:
Forward cmap through geometry guess

mads-bertelsen-agentic committed:
Improve geometry guess failure messages

mads-bertelsen-agentic committed:
Respect verbosity for unknown drawcalls

mads-bertelsen-agentic committed:
Update geometry documentation

mads-bertelsen-agentic committed:
Update geometry test for index bounds

mads-bertelsen-agentic committed:
Keep quiet geometry guess output concise

mads-bertelsen committed:
Merge pull request #127 from mads-bertelsen-agentic/documentation_update

Documentation update and fixes in geometry_viewer module

mads-bertelsen committed:
Version bump and correcting an error message on lack of required modules.

mads-bertelsen committed:
Version bump

Commit:
Update geometry viewer dependency error test

mads-bertelsen committed:
Merge pull request #129 from mads-bertelsen-agentic/fix_tests

Update geometry viewer dependency error test

mads-bertelsen committed:
removed stalte tutorial folder, exists in docs too

mads-bertelsen committed:
Merge pull request #130 from PaNOSC-ViNYL/remove_stale_docs

removed stalte tutorial folder, exists in docs too

mads-bertelsen committed:
lower mpi in docs to ensure run

mads-bertelsen committed:
version bump

mads-bertelsen committed:
Update README.md

mads-bertelsen-agentic committed:
Separate optional dependency tests

mads-bertelsen-agentic committed:
Split geometry viewer tests by dependency

mads-bertelsen-agentic committed:
Declare runtime dependencies in setup

mads-bertelsen-agentic committed:
Consolidate optional renderer imports

mads-bertelsen committed:
Merge pull request #10 from mads-bertelsen-agentic/extra_dependency_tests

Separate optional dependency tests

mads-bertelsen committed:
Merge pull request #132 from mads-bertelsen-agentic/master

Test folders based on dependencies

Review conducted on fork

mads-bertelsen committed:
version bump before release
